edgelet/edgelet-http-workload/src/server/cert/identity.rs

```diff
@@ -1,19 +1,656 @@
+            .name("name")
+            .ok_or_else(|| Error::from(ErrorKind::BadParam))
+            .map(|module_id| {
+                let cn = module_id.to_string();
+                let alias = format!("{}identity", module_id);
+                let module_uri =
+                    prepare_cert_uri_module(cfg.iot_hub_name(), cfg.device_id(), module_id);
+                let result = req
+                    .into_body()
+                    .concat2()
+                    .map(move |body| {
+                        serde_json::from_slice::<IdentityCertificateRequest>(&body)
+                            .context(ErrorKind::BadBody)
+                            .map_err(Error::from)
+                            .and_then(|cert_req| {
+                                match cert_req.expiration() {
+                                    None => Ok(max_duration),
+                                    Some(exp) => compute_validity(exp, max_duration),
+                                }.map(|expiration| expiration)
+                                .map_err(Error::from)
+                            }).and_then(move |expiration| {
+                                hsm.destroy_certificate(alias.clone())
+                                    .map_err(Error::from)?;
+                                let sans = vec![module_uri];
+                                let props = CertificateProperties::new(
+                                    ensure_range!(expiration, 0, max_duration) as u64,
+                                    ensure_not_empty!(cn),
+                                    CertificateType::Client,
+                                    alias,
+                                ).with_san_entries(sans);
+                                hsm.create_certificate(&props)
+                                    .map_err(Error::from)
+                                    .and_then(|cert| {
+                                        let cert = cert_to_response(&cert)?;
+                                        let body = serde_json::to_string(&cert)?;
+                                        Response::builder()
+                                            .status(StatusCode::CREATED)
+                                            .header(CONTENT_TYPE, "application/json")
+                                            .header(CONTENT_LENGTH, body.len().to_string().as_str())
+                                            .body(body.into())
+                                            .map_err(From::from)
+                                    })
+                            }).unwrap_or_else(|e| e.into_response())
+                    }).map_err(Error::from)
+                    .or_else(|e| future::ok(e.into_response()));
+
+                future::Either::A(result)
+            }).unwrap_or_else(|e| future::Either::B(future::ok(e.into_response())));
+
+        Box::new(response)
+    }
+}
+
+fn cert_to_response<T: Certificate>(cert: &T) -> Result<CertificateResponse> {
```

Comment:
This function seems to be basically the same from `./server.rs`. Could we potentially reuse that then?

Reply:
Will refactor this, thanks for pointing this out.